snuba/admin/views.py

```diff
@@ -441,8 +442,65 @@ def clickhouse_trace_query() -> Response:
         )
 
     try:
-        result = run_query_and_get_trace(storage, raw_sql)
-        return make_response(jsonify(asdict(result)), 200)
+        # Append 'log_profile_events=1' with/without the SETTINGS clause to the incoming query.
+        settings_index = raw_sql.lower().find("settings")
+        if settings_index != -1:
+            start_index = settings_index + len("settings")
+            remaining = raw_sql[start_index:].strip()
+            if "log_profile_events=1" not in remaining:
+                raw_sql += ", log_profile_events=1"
+        else:
+            raw_sql += " SETTINGS log_profile_events=1"
+
+        query_trace = run_query_and_get_trace(storage, raw_sql)
+
+        profile_events_raw_sql = "SELECT ProfileEvents FROM system.query_log WHERE query_id = '{}' AND type = 'QueryFinish'"
+        hosts_ports_query_ids_nodes = parse_trace_for_query_ids(query_trace, storage)
+
+        for host_port_query_id_node in hosts_ports_query_ids_nodes:
+            sql = profile_events_raw_sql.format(host_port_query_id_node["query_id"])
+            system_query_result, counter = None, 0
+
+            while counter < 10:
+                # There is a race between the original query to trace finishing and the system query log populating.
+                # Sleep between the query executions.
+                system_query_result = run_system_query_on_host_with_sql(
+                    host_port_query_id_node["host"],
+                    int(host_port_query_id_node["port"]),
+                    storage,
+                    sql,
+                )
+                if not system_query_result.results:
+                    time.sleep(1)
+                    counter += 1
+                else:
+                    break
+
+            if system_query_result is None or len(system_query_result.results) == 0:
+                raise ClickHouseTimeoutError(
+                    "Timed out while executing: {}".format(sql)
+                )
+
+            query_trace.profile_events_meta.append(system_query_result.meta)
+            query_trace.profile_events_profile = cast(
+                Dict[str, int], system_query_result.profile
+            )
+            columns = system_query_result.meta
+            if columns:
+                res = {}
+                res["column_names"] = [name for name, _ in columns]
+                res["rows"] = []
+                for query_result in system_query_result.results:
+                    if query_result[0]:
+                        res["rows"].append(
+                            json.dumps(query_result[0])
+                            .replace("{", "")
+                            .replace("}", "")
```

Comment:
this seems a little odd to me. We want to use `json.dumps` to generate a string that ultimately isn't JSON?

Reply:
Thanks for catching this. I was probably mucking around while passing a string to the admin frontend and make that parsing work and this hack remained in there. I removed it now.